File: sdk/src/realtime.ts
import type { OwnFirebaseConfig } from './types';

export interface RealtimeListenerOptions {
  path: string;
  query?: Record<string, unknown>;
}

export interface RealtimeChange {
  subscriptionId: string;
  event: 'added' | 'modified' | 'removed';
  data: Record<string, unknown>;
  version?: number;
}

export interface RealtimeSnapshot {
  subscriptionId: string;
  data: Record<string, unknown>;
  version: number;
}

export type ChangeListener = (change: RealtimeChange) => void;
export type SnapshotListener = (snapshot: RealtimeSnapshot) => void;
export type ErrorListener = (error: Error) => void;

export class RealtimeSDK {
  private baseUrl: string;
  private projectId: string | undefined;
  private accessToken: string | undefined;
  private ws: WebSocket | null = null;
  private subscriptions = new Map<string, {
    listeners: ChangeListener[];
    snapshotListeners: SnapshotListener[];
    errorListeners: ErrorListener[];
  }>();
  private requestId = 0;
  private pendingRequests = new Map<string, {
    resolve: (value: any) => void;
    reject: (reason?: any) => void;
    timeout: NodeJS.Timeout;
  }>();
  private reconnectAttempts = 0;
  private maxReconnectAttempts = 10;
  private reconnectDelay = 1000;
  private isConnecting = false;

  constructor(config: OwnFirebaseConfig) {
    this.baseUrl = config.baseUrl.replace(/\/$/, '').replace(/^http/, 'ws');
    this.projectId = config.projectId;
    this.accessToken = config.accessToken;
  }

  setAccessToken(token: string): void {
    this.accessToken = token;
  }

  setProjectId(id: string): void {
    this.projectId = id;
  }

  /**
   * Connect to realtime WebSocket. Automatically reconnects on disconnect.
   */
  async connect(): Promise<void> {
    if (this.isConnecting || this.ws?.readyState === WebSocket.OPEN) {
      return;
    }

    this.isConnecting = true;

    return new Promise((resolve, reject) => {
      try {
        const wsUrl = `${this.baseUrl}/ws/v1/projects/${this.projectId}/listen/`;
        this.ws = new WebSocket(wsUrl);

        this.ws.onopen = () => {
          this.isConnecting = false;
          this.reconnectAttempts = 0;
          resolve();
        };

        this.ws.onmessage = (event) => this.handleMessage(event.data);

        this.ws.onerror = (event) => {
          this.isConnecting = false;
          reject(new Error(`WebSocket error: ${event}`));
        };

        this.ws.onclose = () => {
          this.isConnecting = false;
          this.attemptReconnect();
        };
      } catch (error) {
        this.isConnecting = false;
        reject(error);
      }
    });
  }

  /**
   * Disconnect from realtime WebSocket.
   */
  disconnect(): void {
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
    // Clear pending requests
    for (const { timeout } of this.pendingRequests.values()) {
      clearTimeout(timeout);
    }
    this.pendingRequests.clear();
  }

  /**
   * Subscribe to changes on a collection or document path.
   * @example
   * ```ts
   * const unsub = realtime.onSnapshot(
   *   { path: 'users' },
   *   (change) => console.log('Change:', change),
   *   (error) => console.error('Error:', error)
   * );
   * // Later: unsub();
   * ```
   */
  onSnapshot(
    options: RealtimeListenerOptions,
    onNext?: SnapshotListener,
    onError?: ErrorListener
  ): () => void {
    const listener = {
      listeners: [],
      snapshotListeners: onNext ? [onNext] : [],
      errorListeners: onError ? [onError] : [],
    };

    const subscribe = async () => {
      if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
        await this.connect();
      }

      const requestId = String(++this.requestId);
      const subscriptionPromise = this.createPendingRequest(requestId);

      this.ws!.send(JSON.stringify({
        type: 'subscribe',
        requestId,
        path: options.path,
        query: options.query,
      }));

      try {
        const response = await subscriptionPromise;
        const subId = response.subscriptionId;

        if (!this.subscriptions.has(subId)) {
          this.subscriptions.set(subId, listener);
        } else {
          const existing = this.subscriptions.get(subId)!;
          existing.snapshotListeners.push(...listener.snapshotListeners);
          existing.errorListeners.push(...listener.errorListeners);
        }

        // Emit initial snapshot
        if (onNext && response.snapshot) {
          onNext({
            subscriptionId: subId,
            data: response.snapshot,
            version: response.version,
          });
        }

        return subId;
      } catch (error) {
        if (onError) {
          onError(error instanceof Error ? error : new Error(String(error)));
        }
        throw error;
      }
    };

    let subId: string | null = null;
    subscribe().then(id => { subId = id; }).catch(err => {
      if (onError) onError(err);
    });

    // Return unsubscribe function
    return () => {
      if (subId) {
        this.unsubscribe(subId);
      }
    };
  }

  /**
   * Listen to changes on a subscription (after onSnapshot).
   */
  onChange(
    subscriptionId: string,
    onNext: ChangeListener,
    onError?: ErrorListener
  ): () => void {
    if (!this.subscriptions.has(subscriptionId)) {
      this.subscriptions.set(subscriptionId, {
        listeners: [],
        snapshotListeners: [],
        errorListeners: [],
      });
    }

    const sub = this.subscriptions.get(subscriptionId)!;
    sub.listeners.push(onNext);
    if (onError) {
      sub.errorListeners.push(onError);
    }

    // Return unsubscribe function
    return () => {
      const idx = sub.listeners.indexOf(onNext);
      if (idx >= 0) sub.listeners.splice(idx, 1);
      if (onError) {
        const errIdx = sub.errorListeners.indexOf(onError);
        if (errIdx >= 0) sub.errorListeners.splice(errIdx, 1);
      }
    };
  }

  /**
   * Unsubscribe from a realtime listener.
   */
  async unsubscribe(subscriptionId: string): Promise<void> {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
      return;
    }

    const requestId = String(++this.requestId);
    const responsePromise = this.createPendingRequest(requestId, 5000);

    this.ws.send(JSON.stringify({
      type: 'unsubscribe',
      requestId,
      subscriptionId,
    }));

    try {
      await responsePromise;
      this.subscriptions.delete(subscriptionId);
    } catch (error) {
      // Silently fail — just remove subscription
      this.subscriptions.delete(subscriptionId);
    }
  }

  /**
   * Send a ping/pong to keep connection alive.
   */
  ping(): void {
    if (this.ws && this.ws.readyState === WebSocket.OPEN) {
      this.ws.send(JSON.stringify({ type: 'ping' }));
    }
  }

  private handleMessage(data: string): void {
    try {
      const message = JSON.parse(data);

      // Handle pong
      if (message.type === 'pong') {
        return;
      }

      // Handle subscription response
      if (message.type === 'subscribed' && message.requestId) {
        const pending = this.pendingRequests.get(message.requestId);
        if (pending) {
          pending.resolve(message);
          this.pendingRequests.delete(message.requestId);
        }
        return;
      }

      // Handle unsubscribe response
      if (message.type === 'unsubscribed' && message.requestId) {
        const pending = this.pendingRequests.get(message.requestId);
        if (pending) {
          pending.resolve(message);
          this.pendingRequests.delete(message.requestId);
        }
        return;
      }

      // Handle errors
      if (message.type === 'error' && message.requestId) {
        const pending = this.pendingRequests.get(message.requestId);
        if (pending) {
          pending.reject(new Error(`${message.code}: ${message.message}`));
          this.pendingRequests.delete(message.requestId);
        }

        // Also notify error listeners
        if (message.subscriptionId) {
          const sub = this.subscriptions.get(message.subscriptionId);
          if (sub) {
            const error = new Error(`${message.code}: ${message.message}`);
            sub.errorListeners.forEach(listener => listener(error));
          }
        }
        return;
      }

      // Handle data changes
      if (message.type === 'change' && message.subscriptionId) {
        const sub = this.subscriptions.get(message.subscriptionId);
        if (sub) {
          const change: RealtimeChange = {
            subscriptionId: message.subscriptionId,
            event: message.event,
            data: message.data,
            version: message.version,
          };
          sub.listeners.forEach(listener => listener(change));
        }
      }
    } catch (error) {
      console.error('Failed to handle realtime message:', error);
    }
  }

  private createPendingRequest(
    requestId: string,
    timeoutMs = 10000
  ): Promise<any> {
    return new Promise((resolve, reject) => {
      const timeout = setTimeout(() => {
        this.pendingRequests.delete(requestId);
        reject(new Error(`Request ${requestId} timed out`));
      }, timeoutMs);

      this.pendingRequests.set(requestId, { resolve, reject, timeout });
    });
  }

  private attemptReconnect(): void {
    if (this.reconnectAttempts >= this.maxReconnectAttempts) {
      console.error('Max reconnection attempts reached');
      return;
    }

    this.reconnectAttempts++;
    const delay = this.reconnectDelay * Math.pow(2, this.reconnectAttempts - 1);
    const jitteredDelay = delay + (Math.random() - 0.5) * delay * 0.1;

    setTimeout(() => {
      this.connect().catch(error => {
        console.error('Reconnection failed:', error);
      });
    }, jitteredDelay);
  }

  /**
   * Check if WebSocket is currently connected.
   */
  isConnected(): boolean {
    return this.ws?.readyState === WebSocket.OPEN;
  }
}

File: sdk/dart-sdk/lib/realtime.dart
import 'dart:async';
import 'dart:convert';
import 'dart:math' as math;

import 'package:web_socket_channel/web_socket_channel.dart';

import 'types.dart';

/// A single realtime change event delivered for an active subscription.
///
/// Mirrors the server's `{"type": "change", ...}` message. `onChange()`
/// streams the raw decoded message as a `Map<String, dynamic>`; this class
/// is provided as a convenience for callers who want a typed view of that
/// same payload (e.g. `RealtimeChange.fromJson(message)`).
class RealtimeChange {
  final String subscriptionId;
  final String event; // 'added' | 'modified' | 'removed'
  final Map<String, dynamic> data;
  final int? version;
  final String? docId;

  RealtimeChange({
    required this.subscriptionId,
    required this.event,
    required this.data,
    this.version,
    this.docId,
  });

  factory RealtimeChange.fromJson(Map<String, dynamic> json) {
    return RealtimeChange(
      subscriptionId: json['subscriptionId'] as String,
      event: json['event'] as String,
      data: Map<String, dynamic>.from(json['data'] as Map? ?? {}),
      version: json['version'] as int?,
      docId: json['docId'] as String?,
    );
  }

  Map<String, dynamic> toJson() => {
        'type': 'change',
        'subscriptionId': subscriptionId,
        'event': event,
        'data': data,
        'version': version,
        'docId': docId,
      };
}

/// Error returned by the realtime server in response to a request (e.g. a
/// `subscribe` that failed with `PERMISSION_DENIED` or `NOT_FOUND`), or
/// raised locally when a request cannot be completed (timeout, disconnect
/// before a response arrived).
class RealtimeException implements Exception {
  final String code;
  final String message;

  RealtimeException({required this.code, required this.message});

  @override
  String toString() => 'RealtimeException($code): $message';
}

class _PendingRequest {
  final Completer<Map<String, dynamic>> completer;
  final Timer timer;

  _PendingRequest(this.completer, this.timer);
}

/// Realtime (WebSocket) SDK for OwnFirebase.
///
/// Unlike the REST-backed services (`PushSDK`, `DataSDK`, ...), this does not
/// extend `OwnFirebaseClient` since it speaks a JSON-over-WebSocket protocol
/// rather than HTTP. It still takes the same [OwnFirebaseConfig] and exposes
/// [setAccessToken]/[setProjectId] so it can be wired into the SDK bundle the
/// same way the other services are.
///
/// Connects to `ws(s)://<host>/ws/v1/projects/{projectId}/listen/?token=...`.
/// The access token is sent as a query parameter because non-browser clients
/// have no session cookie for the server's WebSocket auth middleware to fall
/// back on.
class RealtimeSDK {
  final String baseUrl;
  String? projectId;
  String? accessToken;

  WebSocketChannel? _channel;
  StreamSubscription<dynamic>? _streamSubscription;
  Future<void>? _connectingFuture;
  bool _connected = false;
  bool _manualDisconnect = false;

  int _requestIdCounter = 0;
  final Map<String, _PendingRequest> _pendingRequests = {};
  final Map<String, StreamController<Map<String, dynamic>>> _changeControllers = {};

  int _reconnectAttempts = 0;
  Timer? _reconnectTimer;
  final math.Random _random = math.Random();

  static const int _reconnectBaseDelayMs = 1000;
  static const int _maxReconnectAttempts = 10;
  static const Duration _subscribeTimeout = Duration(seconds: 10);

  RealtimeSDK({required OwnFirebaseConfig config})
      : baseUrl = config.baseUrl.replaceAll(RegExp(r'/$'), ''),
        projectId = config.projectId,
        accessToken = config.accessToken;

  void setAccessToken(String token) {
    accessToken = token;
  }

  void setProjectId(String id) {
    projectId = id;
  }

  /// Whether the WebSocket is currently open.
  bool get isConnected => _connected;

  // ─── Connection ──────────────────────────────────────────────────────────────

  /// Opens the WebSocket connection. Completes once the connection is open,
  /// or throws if it could not be established. Safe to call when already
  /// connected (no-op) or while a connection attempt is already in flight
  /// (awaits the same attempt).
  Future<void> connect() async {
    _manualDisconnect = false;
    _reconnectTimer?.cancel();
    _reconnectTimer = null;

    if (_connected) {
      return;
    }
    if (_connectingFuture != null) {
      return _connectingFuture;
    }

    final future = _doConnect();
    _connectingFuture = future;
    try {
      await future;
      _reconnectAttempts = 0;
    } finally {
      _connectingFuture = null;
    }
  }

  /// Closes the socket and clears any pending requests. Does not close
  /// existing `onChange` streams — call [unsubscribe] for those.
  void disconnect() {
    _manualDisconnect = true;
    _reconnectTimer?.cancel();
    _reconnectTimer = null;
    _reconnectAttempts = 0;

    _streamSubscription?.cancel();
    _streamSubscription = null;
    _channel?.sink.close();
    _channel = null;
    _connected = false;

    _failAllPending(
      RealtimeException(code: 'INTERNAL', message: 'Disconnected before a response was received'),
    );
  }

  // ─── Subscriptions ───────────────────────────────────────────────────────────

  /// Subscribes to a document or collection path, auto-connecting first if
  /// necessary. Resolves with the server-assigned `subscriptionId` once the
  /// server responds with a `subscribed` acknowledgement (or throws a
  /// [RealtimeException]/[TimeoutException] if it responds with an error or
  /// never responds within 10 seconds).
  Future<String> subscribe(String path, {Map<String, dynamic>? query}) async {
    if (!isConnected) {
      await connect();
    }

    final requestId = _nextRequestId();
    final responseFuture = _awaitResponse(requestId, _subscribeTimeout);

    _send({
      'type': 'subscribe',
      'requestId': requestId,
      'path': path,
      if (query != null) 'query': query,
    });

    final response = await responseFuture;
    final subscriptionId = response['subscriptionId'] as String;
    _changeControllers.putIfAbsent(
      subscriptionId,
      () => StreamController<Map<String, dynamic>>.broadcast(),
    );
    return subscriptionId;
  }

  /// A broadcast stream of `change` messages for [subscriptionId]. Each
  /// event is the raw decoded server message:
  /// `{"type": "change", "subscriptionId": ..., "event": "added"|"modified"|"removed",
  /// "data": {...}, "version": int, "docId": str}`.
  ///
  /// Safe to call before [subscribe] resolves; the underlying controller is
  /// created lazily and shared between both call sites. Cancel the returned
  /// stream's subscription (`.listen(...).cancel()`) to stop listening, or
  /// call [unsubscribe] to also tear down the subscription server-side.
  Stream<Map<String, dynamic>> onChange(String subscriptionId) {
    final controller = _changeControllers.putIfAbsent(
      subscriptionId,
      () => StreamController<Map<String, dynamic>>.broadcast(),
    );
    return controller.stream;
  }

  /// Sends an `unsubscribe` message (if connected) and closes the local
  /// `onChange` stream for [subscriptionId].
  ///
  /// Note: the server protocol defines no success acknowledgement for
  /// `unsubscribe` (only `subscribe` gets a `subscribed` ack) — an `error`
  /// may still arrive asynchronously (e.g. for an unknown subscriptionId),
  /// but since there's nothing actionable to do about it after the fact,
  /// this does not wait for one. The local stream is always torn down.
  Future<void> unsubscribe(String subscriptionId) async {
    if (isConnected) {
      _send({
        'type': 'unsubscribe',
        'requestId': _nextRequestId(),
        'subscriptionId': subscriptionId,
      });
    }

    final controller = _changeControllers.remove(subscriptionId);
    if (controller != null && !controller.isClosed) {
      await controller.close();
    }
  }

  /// Sends a `ping` to keep the connection alive. No-op if not connected.
  void ping() {
    if (isConnected) {
      _send({'type': 'ping'});
    }
  }

  // ─── Internals ───────────────────────────────────────────────────────────────

  Future<void> _doConnect() async {
    final uri = _buildUri();
    final channel = WebSocketChannel.connect(uri);
    await channel.ready;

    _channel = channel;
    _connected = true;
    _streamSubscription = channel.stream.listen(
      _handleRawMessage,
      onError: (Object _, StackTrace __) {
        // Errors on the stream are followed by onDone, which drives
        // reconnection; nothing else to do with the raw error here.
      },
      onDone: _handleSocketClosed,
      cancelOnError: false,
    );
  }

  void _handleSocketClosed() {
    _connected = false;
    _streamSubscription = null;
    _channel = null;

    _failAllPending(
      RealtimeException(code: 'INTERNAL', message: 'Connection closed before a response was received'),
    );

    _scheduleReconnect();
  }

  void _scheduleReconnect() {
    if (_manualDisconnect) return;
    if (_reconnectAttempts >= _maxReconnectAttempts) return;

    _reconnectAttempts++;
    final baseDelay = _reconnectBaseDelayMs * (1 << (_reconnectAttempts - 1));
    final jitter = baseDelay * 0.1 * (_random.nextDouble() - 0.5);
    final delayMs = (baseDelay + jitter).round();

    _reconnectTimer?.cancel();
    _reconnectTimer = Timer(Duration(milliseconds: delayMs), () async {
      try {
        await _doConnect();
        _reconnectAttempts = 0;
      } catch (_) {
        _scheduleReconnect();
      }
    });
  }

  Uri _buildUri() {
    final pid = projectId;
    if (pid == null) {
      throw Exception('projectId is required to open a realtime connection');
    }
    final wsBase = baseUrl.replaceFirst(RegExp(r'^http'), 'ws');
    final token = accessToken;
    final tokenQuery = token != null ? '?token=${Uri.encodeQueryComponent(token)}' : '';
    return Uri.parse('$wsBase/ws/v1/projects/$pid/listen/$tokenQuery');
  }

  void _send(Map<String, dynamic> message) {
    final channel = _channel;
    if (channel == null) {
      throw Exception('Realtime WebSocket is not connected');
    }
    channel.sink.add(jsonEncode(message));
  }

  String _nextRequestId() {
    _requestIdCounter += 1;
    return 'req_$_requestIdCounter';
  }

  Future<Map<String, dynamic>> _awaitResponse(String requestId, Duration timeout) {
    final completer = Completer<Map<String, dynamic>>();
    final timer = Timer(timeout, () {
      if (_pendingRequests.remove(requestId) != null && !completer.isCompleted) {
        completer.completeError(
          TimeoutException('Realtime request $requestId timed out after ${timeout.inSeconds}s'),
        );
      }
    });
    _pendingRequests[requestId] = _PendingRequest(completer, timer);
    return completer.future;
  }

  void _failAllPending(RealtimeException error) {
    for (final pending in _pendingRequests.values) {
      pending.timer.cancel();
      if (!pending.completer.isCompleted) {
        pending.completer.completeError(error);
      }
    }
    _pendingRequests.clear();
  }

  void _handleRawMessage(dynamic raw) {
    if (raw is! String) return;

    Map<String, dynamic> message;
    try {
      final decoded = jsonDecode(raw);
      if (decoded is! Map<String, dynamic>) return;
      message = decoded;
    } catch (_) {
      return;
    }

    switch (message['type'] as String?) {
      case 'pong':
        break;
      case 'subscribed':
        _resolvePending(message['requestId'] as String?, message);
        break;
      case 'error':
        _rejectPending(
          message['requestId'] as String?,
          RealtimeException(
            code: message['code'] as String? ?? 'INTERNAL',
            message: message['message'] as String? ?? 'Unknown realtime error',
          ),
        );
        break;
      case 'change':
        _dispatchChange(message);
        break;
      default:
        break;
    }
  }

  void _resolvePending(String? requestId, Map<String, dynamic> message) {
    if (requestId == null) return;
    final pending = _pendingRequests.remove(requestId);
    if (pending == null) return;
    pending.timer.cancel();
    if (!pending.completer.isCompleted) {
      pending.completer.complete(message);
    }
  }

  void _rejectPending(String? requestId, RealtimeException error) {
    if (requestId == null) return;
    final pending = _pendingRequests.remove(requestId);
    if (pending == null) return;
    pending.timer.cancel();
    if (!pending.completer.isCompleted) {
      pending.completer.completeError(error);
    }
  }

  void _dispatchChange(Map<String, dynamic> message) {
    final subscriptionId = message['subscriptionId'] as String?;
    if (subscriptionId == null) return;
    final controller = _changeControllers[subscriptionId];
    if (controller == null || controller.isClosed) return;
    controller.add(message);
  }
}
